### PR TITLE
Added restrictions around calling MSCCL++ collectives

### PR DESCRIPTION
**Pull request by corey-derochie-amd**
## Details
___Do not mention proprietary info or link to internal work items in this PR.___

**Work item:** Internal

**What were the changes?**  
Applied the same restriction on message sizes to MSCCL++ AllGather as for MSCCL++ AllReduce.

**Why were the changes made?**  
MSCCL++ was being used for invalid message sizes in certain unit tests, causing failures.

## Approval Checklist
___Do not approve until these items are satisfied.___
- [ ] Verify the CHANGELOG has been updated, if
  - there are any NCCL API version changes,
  - any changes impact library users, and/or
  - any changes impact any other ROCm library.

### CODE REVIEW

**Comment by corey-derochie-amd** [comment]
Putting into Ready for Review despite CI failures. I think they need to be re-run. There was a docker issue.

File: cmake/MSCCLPP.cmake [hunk 1]
```diff
@@ -46,7 +46,7 @@ if(ENABLE_MSCCLPP)
     
         download_project(PROJ                mscclpp_nccl
                          GIT_REPOSITORY      https://github.com/microsoft/mscclpp.git
-                         GIT_TAG             b1b9d0626cfa40319c18c05f8c16650568395c29
+                         GIT_TAG             8c6fb429e92e07acb82c0fdcdab44854fc63aa68
```
Comment (nusislam) [inline cmake/MSCCLPP.cmake:49]: Has the bf16 PR been merged to MSCCLPP `main` branch?

File: README.md [hunk 2]
```diff
@@ -134,6 +134,12 @@ MSCCL uses XMLs for different collective algorithms on different architectures.
 
 On the other hand, RCCL allreduce and allgather collectives can leverage the efficient MSCCL++ communication kernels for certain message sizes. MSCCL++ support is available whenever MSCCL support is available. Users need to set the RCCL environment variable `RCCL_ENABLE_MSCCLPP=1` to run RCCL workload with MSCCL++ support. It is also possible to set the message size threshold for using MSCCL++ by using the environment variable `RCCL_MSCCLPP_THRESHOLD`. Once `RCCL_MSCCLPP_THRESHOLD` (the default value is 1MB) is set, RCCL will invoke MSCCL++ kernels for all message sizes less than or equal to the specified threshold.
 
+If some restrictions are not met, it will fall back to MSCCL or RCCL. The following are restrictions on using MSCCL++:
+- Message size must be a non-zero multiple of 32 bytes
+- Does not support `hipMallocManaged` buffers
+- Does not support `ncclBfloat16` or `ncclBfloat8`
```
Comment (nusislam) [inline README.md]: `bf16` should be supported now.

File: cmake/MSCCLPP.cmake (hunk 1 above)
Comment (corey-derochie-amd) [reply cmake/MSCCLPP.cmake:49]: Yes

File: src/misc/msccl/msccl_lifecycle.cc [hunk 3]
```diff
@@ -451,6 +451,28 @@ static ncclResult_t mscclFallBackSavedParams() {
...
+  case ncclFloat16:
+  case ncclInt32:
+  case ncclUint32:
+  case ncclFloat32:
+  case ncclInt64:
+  case ncclUint64:
+  case ncclFloat64:
+#ifdef RCCL_BFLOAT16
+  case ncclBfloat16:
+#endif
```
Comment (corey-derochie-amd) [inline src/misc/msccl/msccl_lifecycle.cc]: Only these are supported, see https://github.com/microsoft/mscclpp/blob/main/apps/nccl/src/nccl.cu#L497
```suggestion
  case ncclFloat16:
  case ncclInt32:
  case ncclUint32:
  case ncclFloat32:
#ifdef RCCL_BFLOAT16
  case ncclBfloat16:
#endif
```

Comment (corey-derochie-amd) [reply src/misc/msccl/msccl_lifecycle.cc]: @nusislam mscclpp doesn't have the same restriction on data type for `ncclAllGather`. Should I reflect that here, or can I keep this global restriction?

File: README.md [hunk 4]
```diff
@@ -134,6 +134,12 @@ MSCCL uses XMLs for different collective algorithms on different architectures.
 
 On the other hand, RCCL allreduce and allgather collectives can leverage the efficient MSCCL++ communication kernels for certain message sizes. MSCCL++ support is available whenever MSCCL support is available. Users need to set the RCCL environment variable `RCCL_ENABLE_MSCCLPP=1` to run RCCL workload with MSCCL++ support. It is also possible to set the message size threshold for using MSCCL++ by using the environment variable `RCCL_MSCCLPP_THRESHOLD`. Once `RCCL_MSCCLPP_THRESHOLD` (the default value is 1MB) is set, RCCL will invoke MSCCL++ kernels for all message sizes less than or equal to the specified threshold.
 
+If some restrictions are not met, it will fall back to MSCCL or RCCL. The following are restrictions on using MSCCL++:
+- Message size must be a non-zero multiple of 32 bytes
+- Does not support `hipMallocManaged` buffers
+- Does not support `ncclBfloat8`
```
Comment (corey-derochie-amd) [inline README.md]: Look into this one.

File: src/misc/msccl/msccl_lifecycle.cc (hunk 3 above)
Comment (corey-derochie-amd) [reply src/misc/msccl/msccl_lifecycle.cc]: Update docs with these missing types.

File: src/misc/msccl/msccl_lifecycle.cc [hunk 5]
```diff
@@ -476,8 +498,13 @@ ncclResult_t mscclEnqueueCheck(
         }
 
         /* check if one rank per GPU and graph mode is enabled */
-        if ((threadLocalStatus.captureStatus != mscclNoCapture) && comm->mscclCompatible) {
-          if (func == mscclFuncAllReduce && nBytes <= comm->mscclpp_threshold && (nBytes & 31) == 0) {
+        if ((threadLocalStatus.captureStatus != mscclNoCapture) && comm->mscclCompatible && isMscclppSupportedType(dataType) && nBytes > 0 && (nBytes & 31) == 0) {
```
Comment (corey-derochie-amd) [inline src/misc/msccl/msccl_lifecycle.cc]: Move the `isMscclppSupportedType` condition down to all reduce only.

**Comment by nusislam** [comment]
LGTM. Did e2e test work with this PR?

**Comment by corey-derochie-amd** [comment]
> LGTM. Did e2e test work with this PR?

It worked in a ROCm 6.1.2 docker using the entire POC branch, which contains this change, but we're still evaluating for perf regressions.

**Pull request merged**